cmem/encoder.go

```diff
@@ -0,0 +1,171 @@
+// Copyright ©2018 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package cmem
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"reflect"
+	"unsafe"
+)
+
+var nativeEndian binary.ByteOrder
+
+func init() {
+	one := uint16(0x1)
+	nativeEndian = [...]binary.ByteOrder{
+		binary.BigEndian,
+		binary.LittleEndian,
+	}[(*[2]byte)(unsafe.Pointer(&one))[0]]
+}
+
+// Encoder is a wrapper type for information necessary to create and
+// subsequently write an in memory object to a PacketTable using Append().
+type Encoder struct {
+	Buf    []byte
+	Offset int
+}
+
+// Encode encodes the value passed as data to binary form stored in []Buf. This
+// buffer is a Go representation of a C in-memory object that can be appended
+// to e.g. a HDF5 PacketTable.
+//
+// Struct values must only have exported fields, otherwise Encode will panic.
+func (enc *Encoder) Encode(data interface{}) error {
+	padding := enc.Offset - len(enc.Buf)
+
+	if padding > 0 {
+		enc.Buf = append(enc.Buf, make([]byte, padding)...)
+	}
+
+	if data, ok := data.(CMarshaler); ok {
+		raw, err := data.MarshalC()
+		if err != nil {
+			return err
+		}
+
+		enc.Buf = append(enc.Buf, raw...)
+
+		return nil
+	}
+
+	rv := reflect.Indirect(reflect.ValueOf(data))
+	rt := rv.Type()
+
+	switch rt.Kind() {
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < rv.Len(); i++ {
+			if err := enc.Encode(rv.Index(i).Interface()); err != nil {
+				return err
+			}
+		}
+		return nil
+
+	case reflect.Struct:
+		offset := enc.Offset
+		for i := 0; i < rv.NumField(); i++ {
+			sfv := rv.Field(i).Interface()
+			// In order to keep the struct offset always correct
+			enc.Offset = offset + int(rt.Field(i).Offset)
+			if err := enc.Encode(sfv); err != nil {
+				return err
+			}
+			// Reset the offset to the correct array sized
+			enc.Offset = offset + int(rt.Size())
+		}
+		return nil
+
+	case reflect.String:
+		str := append([]byte(rv.String()), 0)
+		strPtr := uintptr(unsafe.Pointer(&str[0]))
+		dataSize := int(unsafe.Sizeof(strPtr))
+
+		switch dataSize {
```

Comment:
Interestingly, this can be written

```
var raw [unsafe.Sizeof(strPtr)]byte
switch len(raw) {
case 4:
    nativeEndian.PutUint32(raw[:], uint32(strPtr))
case 8:
    nativeEndian.PutUint64(raw[:], uint64(strPtr))
}

enc.Buf = append(enc.Buf, raw[:]...)
end.Offset += len(raw)
```

`unsafe.Sizeof(T)` is a compiler generated constant, and `len(raw)` is a compile time constant.

Reply:
Nice one!